include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -191,6 +192,334 @@ AffineConstraints<number>::is_consistent_in_parallel(
+        affine_constraints_make_consistent_in_parallel_1;
+
+      // ... determine owners of locally relevant dofs
+      IndexSet locally_relevant_dofs_non_local = locally_relevant_dofs;
+      locally_relevant_dofs_non_local.subtract_set(locally_owned_dofs);
+
+      std::vector<unsigned int> locally_relevant_dofs_owners(
+        locally_relevant_dofs_non_local.n_elements());
+      Utilities::MPI::internal::ComputeIndexOwner::ConsensusAlgorithmsPayload
+        locally_relevant_dofs_process(locally_owned_dofs,
+                                      locally_relevant_dofs_non_local,
+                                      mpi_communicator,
+                                      locally_relevant_dofs_owners,
+                                      true);
+
+      Utilities::MPI::ConsensusAlgorithms::Selector<
+        std::pair<types::global_dof_index, types::global_dof_index>,
+        unsigned int>(locally_relevant_dofs_process, mpi_communicator)
+        .run();
+
+      const auto locally_relevant_dofs_by_ranks =
+        locally_relevant_dofs_process.get_requesters();
+
+      std::map<unsigned int, std::vector<char>> send_data;
+
+      std::vector<MPI_Request> requests;
+      requests.reserve(send_data.size());
+
+      // ... send data
+      for (const auto &rank_and_indices : locally_relevant_dofs_by_ranks)
+        {
+          Assert(rank_and_indices.first != my_rank, ExcInternalError());
+
+          std::vector<ConstraintType> data;
+
+          for (const auto index : rank_and_indices.second)
+            {
+              // note: at this stage locally_relevant_constraints still
+              // contains only locally owned constraints
+              const auto prt =
+                std::find_if(locally_relevant_constraints.begin(),
+                             locally_relevant_constraints.end(),
+                             [index](const auto &a) {
+                               return std::get<0>(a) == index;
+                             });
+              if (prt != locally_relevant_constraints.end())
+                data.push_back(*prt);
+            }
+
+          send_data[rank_and_indices.first] = Utilities::pack(data, false);
+
+          requests.push_back({});
+
+          const int ierr = MPI_Isend(send_data[rank_and_indices.first].data(),
+                                     send_data[rank_and_indices.first].size(),
+                                     MPI_CHAR,
+                                     rank_and_indices.first,
+                                     tag,
+                                     mpi_communicator,
+                                     &requests.back());
+          AssertThrowMPI(ierr);
+        }
+
+      // ... receive data
+      std::set<unsigned int> rec_ranks;
+
+      for (const unsigned int rank : locally_relevant_dofs_owners)
+        if (rank != my_rank)
+          rec_ranks.insert(rank);
+
+      for (unsigned int counter = 0; counter < rec_ranks.size(); ++counter)
```

Comment:
Is this the same as above?

Reply:
No, it is not. `locally_relevant_dofs_owners` stores the rank of each index.

Reply:
I was referring to the fact that you don't need the set and just count (as above), not the code per se :wink:

Reply:
As far as I see it, this is a different station. `locally_relevant_dofs_owners` contains ranks multiple times.

Reply:
I see; I think some comment would help.

Reply:
Next try.